File: Sample/Scripts/Global/Gameplay/ActiveGameplayEffect.lua
local ActiveGameplayEffect = {}

ActiveGameplayEffect.handle = 0
ActiveGameplayEffect.spec = nil
ActiveGameplayEffect.stacks = 1
ActiveGameplayEffect.applicationOrder = 0
ActiveGameplayEffect.grantedAbilitySpecs = {}

function ActiveGameplayEffect:init(handle, spec, applicationOrder)
    self.handle = handle
    self.spec = assert(spec, "Gameplay Effect Spec is required")
    self.stacks = spec.stacks
    self.applicationOrder = applicationOrder
    self.grantedAbilitySpecs = {}
end

return class(ActiveGameplayEffect)

File: Game/Engine/Source/Global/src/Gameplay/ActiveGameplayEffect.cpp
#include <Gameplay/ActiveGameplayEffect.hpp>
#include <Gameplay/GameplayEffectSpec.hpp>

#include <cmath>
#include <stdexcept>
#include <utility>

ActiveGameplayEffect::ActiveGameplayEffect(
    int effectHandle, std::shared_ptr<GameplayEffectSpec> effectSpec, int order)
    : handle(effectHandle),
      spec(std::move(effectSpec)),
      stacks(spec == nullptr ? 1 : spec->stacks),
      applicationOrder(order) {
    if (spec == nullptr) {
        throw std::invalid_argument("Gameplay Effect Spec is required");
    }
}
